net/httpclient.go

```diff
@@ -2,18 +2,114 @@ package net
 
 import (
 	"crypto/tls"
+	"io"
 	"net/http"
 	"net/http/httptrace"
+	"net/url"
+	"strings"
 	"time"
 
 	"github.com/opentracing/opentracing-go"
 	"github.com/opentracing/opentracing-go/ext"
+	"github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/logging"
+	"github.com/zalando/skipper/secrets"
 )
 
 const (
 	defaultIdleConnTimeout = 30 * time.Second
+	defaultRefreshInterval = 5 * time.Minute
 )
 
+type Client struct {
+	client      http.Client
+	log         logging.Logger
+	sr          secrets.SecretsReader
+	secretsPath string
+	quit        chan struct{}
+}
+
+func NewClient(o Options) *Client {
+	quit := make(chan struct{})
+	if o.Log == nil {
+		o.Log = logrus.New()
+	}
+
+	tr := NewTransport(o)
+
+	sr := o.SecretsReader
+	if sr == nil {
+		if o.BearerTokenRefreshInterval == 0 {
+			o.BearerTokenRefreshInterval = defaultRefreshInterval
+		}
+		sp := secrets.NewSecretPaths(o.BearerTokenRefreshInterval)
+		if err := sp.Add(o.BearerTokenFile); err != nil {
+			o.Log.Errorf("failed to read secret: %v", err)
+		}
+		sr = sp
+	}
+
+	c := &Client{
+		client: http.Client{
+			Transport: tr,
+		},
+		log:         o.Log,
+		sr:          sr,
+		secretsPath: o.BearerTokenFile,
+		quit:        quit,
+	}
+
+	return c
+}
+
+func (c *Client) Close() {
+	close(c.quit)
```

Comment:
I don't see c.quit being used, I think we can delete it. OTOH, I think we should call c.client.Transport.Close() somehow (maybe by keeping a reference to our Transport implementation), because NewClient() calls NewTransport() which starts its goroutine.

Reply:
good point, done